dbos/_dbos_config.py

```diff
@@ -323,83 +311,104 @@ def process_config(
-    data["database"]["username"] = (
-        os.getenv("DBOS_DBUSER") or data["database"].get("username") or "postgres"
-    )
-    data["database"]["password"] = (
-        os.getenv("DBOS_DBPASSWORD")
-        or data["database"].get("password")
-        or os.environ.get("PGPASSWORD")
-        or "dbos"
-    )
-
-    if not data["database"].get("app_db_pool_size"):
-        data["database"]["app_db_pool_size"] = 20
-    if not data["database"].get("sys_db_pool_size"):
-        data["database"]["sys_db_pool_size"] = 20
-    if not data["database"].get("connectionTimeoutMillis"):
-        data["database"]["connectionTimeoutMillis"] = 10000
-
+    # Handle admin server config
     if not data.get("runtimeConfig"):
         data["runtimeConfig"] = {
             "run_admin_server": True,
         }
-    elif "run_admin_server" not in data["runtimeConfig"]:
+    elif data["runtimeConfig"].get("run_admin_server") is None:
         data["runtimeConfig"]["run_admin_server"] = True
 
+    isDebugMode = os.getenv("DBOS_DBHOST") is not None
+
+    # Ensure database dict exists
+    data.setdefault("database", {})
+    configure_db_engine_parameters(data["database"])
+
+    # Database URL resolution
+    if data.get("database_url") is not None and data["database_url"] != "":
+        # Parse the db string and check required fields
+        assert data["database_url"] is not None
+        url = make_url(data["database_url"])
+        required_fields = [
+            ("username", "Username must be specified in the connection URL"),
+            ("password", "Password must be specified in the connection URL"),
+            ("host", "Host must be specified in the connection URL"),
+            ("database", "Database name must be specified in the connection URL"),
+        ]
+        for field_name, error_message in required_fields:
+            field_value = getattr(url, field_name, None)
+            if not field_value:
+                raise DBOSInitializationError(error_message)
+
+        if not data["database"].get("sys_db_name"):
+            assert url.database is not None
+            data["database"]["sys_db_name"] = url.database + SystemSchema.sysdb_suffix
+
+        # In debug mode perform env vars overrides
+        if isDebugMode:
+            # Override the username, password, host, and port
+            port_str = os.getenv("DBOS_DBPORT")
+            port = (
+                int(port_str)
+                if port_str is not None and port_str.isdigit()
+                else url.port
+            )
+            data["database_url"] = url.set(
+                username=os.getenv("DBOS_DBUSER", url.username),
+                password=os.getenv("DBOS_DBPASSWORD", url.password),
+                host=os.getenv("DBOS_DBHOST", url.host),
+                port=port,
+            ).render_as_string(hide_password=False)
+    else:
+        _app_db_name = _app_name_to_db_name(data["name"])
+        _password = os.environ.get("PGPASSWORD", "dbos")
```

Comment:
Note we do *not* escape the password here, and expect users to do so. This is to avoid escaping a password twice.